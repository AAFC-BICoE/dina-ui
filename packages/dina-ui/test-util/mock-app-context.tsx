import {
  AccountContextI,
  AccountProvider,
  ApiClientI,
  ApiClientImpl,
  ApiClientProvider,
  ModalProvider
} from "common-ui";
import { mount } from "enzyme";
import { merge, noop } from "lodash";
import { useMemo, useRef } from "react";
import { DndProvider } from "react-dnd-cjs";
import HTML5Backend from "react-dnd-html5-backend-cjs";
import { SWRConfig } from "swr";
import { PartialDeep } from "type-fest";
import { FileUploadProviderImpl } from "../components/object-store/file-upload/FileUploadProvider";
import { DinaIntlProvider } from "../intl/dina-ui-intl";

interface MockAppContextProviderProps {
  apiContext?: PartialDeep<ApiClientI>;
  accountContext?: Partial<AccountContextI>;
  children?: React.ReactNode;
}

/**
 * Wraps a test-rendered component to provide the contexts that would be available in
 * the application.
 */
export function MockAppContextProvider({
  accountContext,
  apiContext = { apiClient: { get: () => undefined as any } },
  children
}: MockAppContextProviderProps) {
  const DEFAULT_MOCK_ACCOUNT_CONTEXT: AccountContextI = useMemo(
    () => ({
      authenticated: true,
      groupNames: ["aafc", "cnc"],
      initialized: true,
      login: noop,
      logout: noop,
      roles: ["user"],
      token: "test-token",
      username: "test-user",
      isAdmin: false
    }),
    []
  );

  const DEFAULT_API_CONTEXT_VALUE = useMemo(
    () =>
      new ApiClientImpl({
        newId: () => "00000000-0000-0000-0000-000000000000"
      }),
    []
  );

  const apiContextWithWarnings = {
    ...apiContext,
    // Add a warning when bulkGet doesn't return anything in a test:
    bulkGet: async (paths: string[], options) => {
      const resources = await apiContext?.bulkGet?.(paths, options);
      if (!resources) {
        console.warn("No response returned for bulkGet paths: ", paths);
      } else {
        resources.forEach((resource, index) => {
          if (!resource) {
            console.warn("No value returned for bulkGet path: ", paths[index]);
          }
        });
      }
      return resources;
    }
  };

  const modalWrapperRef = useRef<HTMLDivElement>(null);

  return (
    <SWRConfig
      // Reset SWR cache between tests.
      value={{ provider: () => new Map() }}
    >
      <AccountProvider
        value={{ ...DEFAULT_MOCK_ACCOUNT_CONTEXT, ...accountContext }}
      >
        <ApiClientProvider
          value={merge({}, DEFAULT_API_CONTEXT_VALUE, apiContextWithWarnings)}
        >
          <FileUploadProviderImpl>
            <DinaIntlProvider>
              <DndProvider backend={HTML5Backend}>
                <div ref={modalWrapperRef}>
                  <ModalProvider appElement={modalWrapperRef.current}>
                    {children}
                  </ModalProvider>
                </div>
              </DndProvider>
            </DinaIntlProvider>
          </FileUploadProviderImpl>
        </ApiClientProvider>
      </AccountProvider>
    </SWRConfig>
  );
}

/**
 * Helper function to get a test wrapper with the required context providers.
 */
export function mountWithAppContext(
  element: React.ReactNode,
  mockAppContextProviderProps?: MockAppContextProviderProps
) {
  return mount(
    <MockAppContextProvider {...mockAppContextProviderProps}>
      {element}
    </MockAppContextProvider>
  );
}
