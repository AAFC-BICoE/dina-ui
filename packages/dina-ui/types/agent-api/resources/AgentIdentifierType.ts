import { MultilingualTitle } from "../../common";

export interface AgentIdentifierTypeAttributes {
  type: "identifier-type";
  id?: string | undefined;
  createdBy?: string;
  createdOn?: string;
  key?: string;
  name?: string;
  dinaComponents?: string[];
  uriTemplate?: string;
  term?: string;
  multilingualTitle?: MultilingualTitle;
}

export type AgentIdentifierType = AgentIdentifierTypeAttributes;
