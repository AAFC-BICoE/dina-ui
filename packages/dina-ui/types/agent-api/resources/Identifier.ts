export interface PersonIdentifierAttributes {
  type: "identifier";
  id?: string | undefined;
  createdBy?: string;
  createdOn?: string;
  namespace?: string;
  value?: string;
}

export type Identifier = PersonIdentifierAttributes;
