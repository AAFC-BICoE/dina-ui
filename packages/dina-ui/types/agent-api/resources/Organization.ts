import { KitsuResource } from "kitsu";

export interface OrganizationAttributes {
  type: "organization";
  name?: Map<string, string>;
  names: MultiligualName[];
  aliases?: string[];
  uuid: string;
  createdBy?: string;
  createdOn?: string;
}

export type MultiligualName = {
  languageCode: string;
  name: string;
};

export type Organization = KitsuResource & OrganizationAttributes;
