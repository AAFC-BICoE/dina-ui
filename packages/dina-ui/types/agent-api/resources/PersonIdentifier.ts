export interface PersonIdentifierAttributes {
  type?: PersonIdentifierType;
  uri?: string;
}

export enum PersonIdentifierType {
  ORCID = "ORCID",
  WIKIDATA = "WIKIDATA"
}

export type PersonIdentifier = PersonIdentifierAttributes;
