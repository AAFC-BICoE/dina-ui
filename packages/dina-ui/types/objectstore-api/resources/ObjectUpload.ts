import { KitsuResource } from "kitsu";

export interface ObjectUploadAttributes {
  fileIdentifier: string;
  metaFileEntryVersion: string;
  originalFilename: string;
  sha1Hex: string;
  receivedMediaType: string;
  detectedMediaType: string;
  detectedFileExtension: string;
  evaluatedMediaType: string;
  evaluatedFileExtension: string;
  sizeInBytes: number;
  exif: Map<string, string>;
  dateTimeDigitized?: string;
}

export type ObjectUpload = KitsuResource & ObjectUploadAttributes;
