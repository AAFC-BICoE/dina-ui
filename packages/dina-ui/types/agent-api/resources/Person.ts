import { KitsuResource } from "kitsu";
import { Identifier } from "./Identifier";
import { Organization } from "./Organization";

export interface PersonAttributes {
  type: "person";
  displayName?: string;
  givenNames?: string;
  familyNames?: string;
  aliases?: string[];
  email?: string;
  uuid?: string | undefined;
  createdBy?: string;
  createdOn?: string;
  webpage?: URL;
  remarks?: string;
}

export interface PersonRelationships {
  organizations?: Organization[];
  identifiers?: Identifier[];
}

export type Person = KitsuResource & PersonAttributes & PersonRelationships;
