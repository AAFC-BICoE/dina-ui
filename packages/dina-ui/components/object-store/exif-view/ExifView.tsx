import {
  ColumnDefinition,
  LoadingSpinner,
  useCollapser,
  useQuery
} from "common-ui";
import ReactTable from "react-table";
import { ReactNode } from "react";
import { ObjectUpload } from "packages/dina-ui/types/objectstore-api/resources/ObjectUpload";
import { useDinaIntl } from "../../../intl/dina-ui-intl";

interface ExifProps {
  exif: Map<string, string>;
}

interface ObjectUploadProps {
  objectUpload: ObjectUpload;
}

function DisplayExif({ exif }: ExifProps) {
  const { formatMessage } = useDinaIntl();
  const ExifMaps = Object.keys(exif).map(key => {
    return {
      propKey: key,
      propValue: exif[key]
    };
  });
  const METADATA_TABLE_COLUMNS = [
    {
      Header: formatMessage("exifAttribute"),
      accessor: "propKey"
    },
    {
      Header: formatMessage("exifValue"),
      accessor: "propValue"
    }
  ];
  return (
    <ReactTable
      className="-striped"
      columns={METADATA_TABLE_COLUMNS}
      showPagination={false}
      data={ExifMaps}
    />
  );
}

interface CollapsableSectionProps {
  children: ReactNode;
  collapserId: string;
  title: ReactNode;
}

/** Wrapper for the collapsible sections of the details UI. */
function CollapsableSection({
  children,
  collapserId,
  title
}: CollapsableSectionProps) {
  const { Collapser, collapsed } = useCollapser(`view-exif-${collapserId}`);

  return (
    <div className="form-group">
      <h4>
        {title}
        <Collapser />
      </h4>
      {!collapsed && children}
    </div>
  );
}
export function ExifView({ objectUpload }: ObjectUploadProps) {
  if (objectUpload && objectUpload.exif) 
  {    
    const { formatMessage } = useDinaIntl();
    return (
      <CollapsableSection
        collapserId={objectUpload.fileIdentifier}
        title={formatMessage("exifProperties")}
        key={objectUpload.fileIdentifier}
      >
      <DisplayExif exif={objectUpload.exif} />
      </CollapsableSection>
    );
  }
  return null;
}
