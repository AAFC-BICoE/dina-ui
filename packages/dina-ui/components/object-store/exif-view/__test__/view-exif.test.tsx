import { AccountContextI } from "common-ui";
import { noop } from "lodash";
import { ObjectUpload } from "../../../../types/objectstore-api/resources/ObjectUpload";
import { mountWithAppContext } from "../../../../test-util/mock-app-context";
import { ExifView } from "../ExifView";

const exifData = new Map().set("date original created", "2000, Jan 8");

/** Test ObjectUpload */
const TEST_OBJECTUPLOAD: ObjectUpload = {
  fileIdentifier: "c0f78fce-1825-4c4e-89c7-92fe0ed9dc73",
  sizeInBytes: 500,
  originalFilename: "test.png",
  metaFileEntryVersion: "1",
  sha1Hex: "da39a3ee5e6b4b0d3255bfef95601890afd80709",
  receivedMediaType: "image/png",
  detectedMediaType: "image/png",
  detectedFileExtension: "png",
  evaluatedMediaType: "image/png",
  evaluatedFileExtension: "png",
  exif: Object.fromEntries(exifData),
  type: "object-upload"
};

const MOCK_ACCOUNT_CONTEXT: AccountContextI = {
  agentId: "6ee06232-e801-4cd5-8fc5-127aa14c3ace",
  authenticated: true,
  groupNames: ["example-group"],
  initialized: true,
  login: noop,
  logout: noop,
  roles: [],
  // Mock for a successful token update.
  getCurrentToken: () => Promise.resolve("test-token"),
  username: "test-user"
};

describe("View Exif page", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("renders exif props when received file upload response", () => {
    const elm = <ExifView objectUpload={TEST_OBJECTUPLOAD} />;
    const wrapper = mountWithAppContext(elm, {
      accountContext: { ...MOCK_ACCOUNT_CONTEXT, groupNames: [] }
    });

    expect(
      wrapper.find("div.key-cell.rt-td").contains("Date Original Created")
    ).toBe(true);
    expect(wrapper.find("div.value-cell.rt-td").contains("2000, Jan 8")).toBe(
      true
    );
  });
});
