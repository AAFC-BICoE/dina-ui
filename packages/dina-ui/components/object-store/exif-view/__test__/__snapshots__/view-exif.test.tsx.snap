// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View Exif page renders exif props when received file upload response 1`] = `
<DocumentFragment>
  <div>
    <div
      class="mb-3"
    >
      <h4>
        Exif Properties
        <button
          aria-expanded="true"
          aria-label="Collapse section"
          class="collapser-button btn btn-light "
          style="display: inline-flex; align-items: center; justify-content: center; width: 2.5rem; height: 2.5rem; margin-top: -5px; margin-left: 10px; background: transparent;"
          title="Collapse section"
          type="button"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            style="color: rgb(0, 0, 0); font-size: 1.5rem; transform: rotate(0deg); transition: transform 2s cubic-bezier(.4,2,.55,.44); will-change: transform;"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M201.4 342.6c12.5 12.5 32.8 12.5 45.3 0l160-160c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 274.7 86.6 137.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l160 160z"
            />
          </svg>
        </button>
      </h4>
      <div
        class="ReactTable -striped"
        data-testid="ReactTable"
      >
        <div
          class="table-responsive-wrapper"
        >
          <table
            class="w-100"
          >
            <thead>
              <tr>
                <th
                  class="-cursor-pointer"
                  colspan="1"
                  style="width: 200px;"
                >
                  <div
                    class="-cursor-pointer select-none column-header"
                  >
                    <span
                      class="d-flex align-items-center justify-content-center"
                    >
                      Attribute
                    </span>
                  </div>
                </th>
                <th
                  class="-cursor-pointer"
                  colspan="1"
                  style="width: auto;"
                >
                  <div
                    class="-cursor-pointer select-none column-header"
                  >
                    <span
                      class="d-flex align-items-center justify-content-center"
                    >
                      Value
                    </span>
                  </div>
                </th>
              </tr>
            </thead>
            <tbody>
              <tr
                class="index-0 -odd"
              >
                <td
                  class="key-cell"
                >
                  <strong>
                    <div
                      class="date original created-field-header"
                    >
                      Date Original Created
                    </div>
                  </strong>
                </td>
                <td
                  class="value-cell"
                >
                  <div
                    class="field-view"
                    style="white-space: pre-wrap;"
                  >
                    2000, Jan 8
                  </div>
                </td>
              </tr>
              <tr
                class="index-1 -even"
              >
                <td
                  class="key-cell"
                >
                  <strong>
                    <div
                      class="shutter speed-field-header"
                    >
                      Shutter Speed
                    </div>
                  </strong>
                </td>
                <td
                  class="value-cell"
                >
                  <div
                    class="field-view"
                    style="white-space: pre-wrap;"
                  >
                    1/200
                  </div>
                </td>
              </tr>
              <tr
                class="index-2 -odd"
              >
                <td
                  class="key-cell"
                >
                  <strong>
                    <div
                      class="exposure-field-header"
                    >
                      Exposure
                    </div>
                  </strong>
                </td>
                <td
                  class="value-cell"
                >
                  <div
                    class="field-view"
                    style="white-space: pre-wrap;"
                  >
                    0.5
                  </div>
                </td>
              </tr>
              <tr
                class="index-3 -even"
              >
                <td
                  class="key-cell"
                >
                  <strong>
                    <div
                      class="compensation-field-header"
                    >
                      Compensation
                    </div>
                  </strong>
                </td>
                <td
                  class="value-cell"
                >
                  <div
                    class="field-view"
                    style="white-space: pre-wrap;"
                  >
                    -0.5
                  </div>
                </td>
              </tr>
              <tr
                class="index-4 -odd"
              >
                <td
                  class="key-cell"
                >
                  <strong>
                    <div
                      class="aperture-field-header"
                    >
                      Aperture
                    </div>
                  </strong>
                </td>
                <td
                  class="value-cell"
                >
                  <div
                    class="field-view"
                    style="white-space: pre-wrap;"
                  >
                    f/8
                  </div>
                </td>
              </tr>
              <tr
                class="index-5 -even"
              >
                <td
                  class="key-cell"
                >
                  <strong>
                    <div
                      class="frame rate-field-header"
                    >
                      Frame Rate
                    </div>
                  </strong>
                </td>
                <td
                  class="value-cell"
                >
                  <div
                    class="field-view"
                    style="white-space: pre-wrap;"
                  >
                    30fps
                  </div>
                </td>
              </tr>
              <tr
                class="index-6 -odd"
              >
                <td
                  class="key-cell"
                >
                  <strong>
                    <div
                      class="f-number-field-header"
                    >
                      F Number
                    </div>
                  </strong>
                </td>
                <td
                  class="value-cell"
                >
                  <div
                    class="field-view"
                    style="white-space: pre-wrap;"
                  >
                    f/2.8
                  </div>
                </td>
              </tr>
              <tr
                class="index-7 -even"
              >
                <td
                  class="key-cell"
                >
                  <strong>
                    <div
                      class="resolution-field-header"
                    >
                      Resolution
                    </div>
                  </strong>
                </td>
                <td
                  class="value-cell"
                >
                  <div
                    class="field-view"
                    style="white-space: pre-wrap;"
                  >
                    300dpi
                  </div>
                </td>
              </tr>
              <tr
                class="index-8 -odd"
              >
                <td
                  class="key-cell"
                >
                  <strong>
                    <div
                      class="self-time mode usage-field-header"
                    >
                      Self Time Mode Usage
                    </div>
                  </strong>
                </td>
                <td
                  class="value-cell"
                >
                  <div
                    class="field-view"
                    style="white-space: pre-wrap;"
                  >
                    10s
                  </div>
                </td>
              </tr>
              <tr
                class="index-9 -even"
              >
                <td
                  class="key-cell"
                >
                  <strong>
                    <div
                      class="camera model-field-header"
                    >
                      Camera Model
                    </div>
                  </strong>
                </td>
                <td
                  class="value-cell"
                >
                  <div
                    class="field-view"
                    style="white-space: pre-wrap;"
                  >
                    Canon EOS R5
                  </div>
                </td>
              </tr>
              <tr
                class="index-10 -odd"
              >
                <td
                  class="key-cell"
                >
                  <strong>
                    <div
                      class="orientation-field-header"
                    >
                      Orientation
                      <span
                        class="m-2"
                      >
                        <span>
                          <img
                            alt="Determines the way an image should be displayed based on its orientation value, must be a number from 1 to 8."
                            aria-describedby="field_orientation_tooltip"
                            src="/static/images/iconInformation.gif"
                            tabindex="0"
                          />
                        </span>
                      </span>
                    </div>
                  </strong>
                </td>
                <td
                  class="value-cell"
                >
                  <div
                    class="field-view"
                    style="white-space: pre-wrap;"
                  >
                    landscape
                  </div>
                </td>
              </tr>
              <tr
                class="index-11 -even"
              >
                <td
                  class="key-cell"
                >
                  <strong>
                    <div
                      class="metering mode-field-header"
                    >
                      Metering Mode
                    </div>
                  </strong>
                </td>
                <td
                  class="value-cell"
                >
                  <div
                    class="field-view"
                    style="white-space: pre-wrap;"
                  >
                    evaluative
                  </div>
                </td>
              </tr>
              <tr
                class="index-12 -odd"
              >
                <td
                  class="key-cell"
                >
                  <strong>
                    <div
                      class="ISO speed information-field-header"
                    >
                      ISO Speed Information
                    </div>
                  </strong>
                </td>
                <td
                  class="value-cell"
                >
                  <div
                    class="field-view"
                    style="white-space: pre-wrap;"
                  >
                    100
                  </div>
                </td>
              </tr>
              <tr
                class="index-13 -even"
              >
                <td
                  class="key-cell"
                >
                  <strong>
                    <div
                      class="pixel dimensions-field-header"
                    >
                      Pixel Dimensions
                    </div>
                  </strong>
                </td>
                <td
                  class="value-cell"
                >
                  <div
                    class="field-view"
                    style="white-space: pre-wrap;"
                  >
                    6720x4480
                  </div>
                </td>
              </tr>
              <tr
                class="index-14 -odd"
              >
                <td
                  class="key-cell"
                >
                  <strong>
                    <div
                      class="copyright information-field-header"
                    >
                      Copyright Information
                    </div>
                  </strong>
                </td>
                <td
                  class="value-cell"
                >
                  <div
                    class="field-view"
                    style="white-space: pre-wrap;"
                  >
                    John Doe
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
